subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/LauncherServices.java

```diff
@@ -69,19 +73,48 @@ public void registerProjectServices(ServiceRegistration registration) {
     }
 
     static class ToolingGlobalScopeServices {
-        BuildExecuter createBuildExecuter(GradleLauncherFactory gradleLauncherFactory, ServiceRegistry globalServices, ListenerManager listenerManager, FileWatcherFactory fileWatcherFactory, ExecutorFactory executorFactory, StyledTextOutputFactory styledTextOutputFactory, GradleUserHomeScopeServiceRegistry userHomeServiceRegistry) {
-            List<BuildActionRunner> buildActionRunners = globalServices.getAll(BuildActionRunner.class);
+
+        // Can't expose SubscribableBuildActionRunner directly as the call to globalServices.getAll(BuildActionRunner.class)
+        // then fail with a StackOverflowError
+        SubscribableBuildActionRunnerProvider createSubscribableBuildActionRunnerProvider(final ServiceRegistry globalServices) {
+            BuildOperationExecutor buildOperationExecutor = globalServices.get(BuildOperationExecutor.class);
+            BuildOperationService buildOperationService = globalServices.get(BuildOperationService.class);
+            Factory<List<? extends BuildActionRunner>> runnersFactory = new Factory<List<? extends BuildActionRunner>>() {
```

Comment:
I think we shouldn't need these two factories. All registration should be done before the first thing gets created.